src/Components/Web/src/Forms/InputRadioGroup.cs

```diff
@@ -30,21 +30,33 @@ public class InputRadioGroup<[DynamicallyAccessedMembers(DynamicallyAccessedMemb
     /// <inheritdoc />
     protected override void OnParametersSet()
     {
-        var groupName = !string.IsNullOrEmpty(Name) ? Name : _defaultGroupName;
-        var fieldClass = EditContext?.FieldCssClass(FieldIdentifier) ?? string.Empty;
-        var changeEventCallback = EventCallback.Factory.CreateBinder<string?>(this, __value => CurrentValueAsString = __value, CurrentValueAsString);
+        // On the first render, we can instantiate the InputRadioContext
+        if (_context is null)
+        {
+            var changeEventCallback = EventCallback.Factory.CreateBinder<string?>(this, __value => CurrentValueAsString = __value, CurrentValueAsString);
+            _context = new InputRadioContext(CascadedContext, changeEventCallback);
+        }
+        else if (_context.ParentContext != CascadedContext)
+        {
+            // This should never be possible in any known usage pattern, but if it happens, we want to know
+            throw new InvalidOperationException("An InputRadioGroup cannot change context after creation");
+        }
 
-        _context = new InputRadioContext(CascadedContext, groupName, CurrentValue, fieldClass, changeEventCallback);
+        // Mutate the InputRadioContext instance in place. Since this is a non-fixed cascading parameter, the descendant
+        // InputRadio/InputRadioGroup components will get notified to re-render and will see the new values.
+        _context.GroupName = !string.IsNullOrEmpty(Name) ? Name : _defaultGroupName;
```

Comment:
Could you walk me thru how this work? Wouldn't this mean the old and new cascading parameter values are the same instance and so are always the same?

Reply:
@pranavkm 
The cascading parameter values are indeed the same, but because of ``IsFixed`` now being set to false, it will have subscribers. 
When the ``SetParametersAsync(ParameterView parameters)`` of the CascadingValue is called the following code is run

https://github.com/dotnet/aspnetcore/blob/6f5aeb71eabbe19155bede07ab029d06a3459d22/src/Components/Components/src/CascadingValue.cs#L125-L128

And due to the fact that ``ChangeDetection.MayHaveChanged`` will return true for any class parameter,  due to ``NotifySubscribers`` it will eventually end up calling ``IComponent.SetParametersAsync(ParameterView parameters)`` on every ``InputRadio`` component, forcing it to rerender.

Reply:
+1 to everything @MariovanZeist said

And to emphasise, this is not some obscure behaviour we’re now relying on, it’s central to how parameter passing and cascaded parameter notifications work. Any possibly-mutable type is always regarded as possibly-changed and hence triggers a re-render.